### ISSUE TITLE
expect_lint() error for missing {testthat} should reference which function called it

### ISSUE THREAD

**Issue by MichaelChirico** [r-lib/lintr#2585]
ideally this will just reference the actual call for clarity, but we can save that for later.

_Originally posted by @MichaelChirico in https://github.com/r-lib/lintr/pull/2582#discussion_r1616310046_

**Comment by F-Noelle** [comment]
I am thinking something along the line of: 
```
who_called_me <- as.character(sys.call(sys.parent())) |>
  stringr::str_split(stringr::fixed("("))
who_called_me <- who_called_me[[1]]
```
I am not sure if there would be a more elegant way to get this info which function did the original call.

Or you create a third unexported function called "expect_lint_logic" who does all the heavy lifting and has an addition parameter which is simply the name of the calling function. 

Any better ideas? 

I wanted to write the expect_no_lint_linter as well, given that this should be a fairly easy linter to write as it just needs to check for NULL on the second positional parameter of the function, as far as I understood the xpath logic

**Comment by MichaelChirico** [comment]
> I wanted to write the expect_no_lint_linter as well, given that this should be a fairly easy linter to write as it just needs to check for NULL on the second positional parameter of the function, as far as I understood the xpath logic

Follow #1874 for some discussion about how exactly to go about that. We might want something more generalizable than just `expect_no_lint_linter()`.

**Comment by AshesITR** [comment]
Regarding the caller logic, you can look at the code of `Linter()` which auto-names using a similar approach.

**Comment by MichaelChirico** [comment]
I would use:

```r
tail(as.character(sys.call(sys.parent())[[1]]), 1L)
```

That covers `lintr::expect_no_lint()` and plain `expect_no_lint()`.

The `lintr:::linter_auto_name()` logic _might_ be overkill in this case.

Alternatively, we can just copy-paste the `requireNamespace()` call into the body of `expect_no_lint()`.

**Comment by AshesITR** [comment]
> Alternatively, we can just copy-paste the `requireNamespace()` call into the body of `expect_no_lint()`.

This is the way to go I think.

**Comment by F-Noelle** [comment]
Probably the most simple solution indeed.
